File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/actor/map/TilesetRequestActor.java
package net.bestia.zoneserver.actor.map;

import akka.actor.ActorRef;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import net.bestia.messages.map.MapTilesetMessage;
import net.bestia.messages.map.MapTilesetRequestMessage;
import net.bestia.model.domain.TilesetData;
import net.bestia.model.map.Tileset;
import net.bestia.model.map.TilesetService;
import net.bestia.zoneserver.actor.SpringExtension;
import net.bestia.zoneserver.actor.routing.BaseClientMessageRouteActor;
import net.bestia.zoneserver.actor.client.SendToClientActor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.util.Objects;
import java.util.Optional;

/**
 * The user queries the name/data of an {@link TilesetData}. He only sends the
 * GID of the tile and we must find the appropriate tileset.
 * 
 * @author Thomas Felix
 *
 */
@Component
@Scope("prototype")
public class TilesetRequestActor extends BaseClientMessageRouteActor {

	private final LoggingAdapter LOG = Logging.getLogger(getContext().system(), this);
	public static String NAME = "tileset";

	private final TilesetService tilesetService;
	private final ActorRef sendClient;

	@Autowired
	public TilesetRequestActor(TilesetService tilesetService) {

		this.tilesetService = Objects.requireNonNull(tilesetService);
		this.sendClient = SpringExtension.actorOf(getContext(), SendToClientActor.class);
		getRedirectConfig().match(MapTilesetRequestMessage.class, this::onMapTilesetRequest);
	}

	private void onMapTilesetRequest(MapTilesetRequestMessage msg) {

		final Optional<Tileset> ts = tilesetService.findTileset(msg.getTileId());

		if (!ts.isPresent()) {
			LOG.warning("Tileset containing gid {} not found.", msg.getTileId());
			return;
		}

		final MapTilesetMessage response = new MapTilesetMessage(
				msg.getAccountId(),
				ts.get().getSimpleTileset());

		sendClient.tell(response, getSelf());
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/actor/map/TilesetRequestActor.kt
package net.bestia.zoneserver.actor.map

import mu.KotlinLogging
import net.bestia.messages.map.MapTilesetMessage
import net.bestia.messages.map.MapTilesetRequestMessage
import net.bestia.model.map.TilesetData
import net.bestia.zoneserver.actor.Actor
import net.bestia.zoneserver.actor.SpringExtension
import net.bestia.zoneserver.actor.client.SendToClientActor
import net.bestia.zoneserver.actor.routing.DynamicMessageRouterActor
import net.bestia.zoneserver.map.TilesetService

private val LOG = KotlinLogging.logger { }

/**
 * The user queries the name/data of an [TilesetData]. He only sends the
 * GID of the tile and we must find the appropriate tileset.
 *
 * @author Thomas Felix
 */
@Actor
class TilesetRequestActor(
    private val tilesetService: TilesetService
) : DynamicMessageRouterActor() {

  private val sendClient = SpringExtension.actorOf(context, SendToClientActor::class.java)

  override fun createReceive(builder: BuilderFacade) {
    builder.match(MapTilesetRequestMessage::class.java, this::onMapTilesetRequest)
  }

  private fun onMapTilesetRequest(msg: MapTilesetRequestMessage) {

    val ts = tilesetService.findTileset(msg.tileId)
    if (ts == null) {
      LOG.warn { "Tileset containing gid ${msg.tileId} not found." }
      return
    }

    val response = MapTilesetMessage(
        msg.accountId,
        ts.toTilesetDTO()
    )

    sendClient.tell(response, self)
  }

  companion object {
    const val NAME = "tileset"
  }
}